File: src/jray/math/Matrix.java
package jray.math;

import jray.common.Matrix4;
import jray.common.Vect3;

public class Matrix {

    /**
     * creates a rotation matrix for affine transformation
     * @param axis axis to rotate
     * @param angleRad angle to rotate
     * @param erg Matrix4 object to store the result
     */
    public static void createRotationMatrix(Vect3 axis, double angleRad, Matrix4 erg) {
        double[][] data = erg.getData();
        double[] axisdata = axis.getData();
        double cosa = Math.cos(angleRad);
        double sina = Math.sin(angleRad);

        double icosa = 1 - cosa; // performance ... every bit matters
        double v1xicosa = axisdata[0] * icosa;

        data[0][0] = cosa + axisdata[0] * v1xicosa;
        data[0][1] = axisdata[1] * v1xicosa - axisdata[2] * sina;
        data[0][2] = axisdata[2] * v1xicosa + axisdata[1] * sina;
        data[0][3] = 0;
        data[1][0] = axisdata[1] * v1xicosa + axisdata[2] * sina;
        data[1][1] = cosa + axisdata[1] * axisdata[1] * icosa;
        data[1][2] = axisdata[1] * axisdata[2] * icosa - axisdata[0] * sina;
        data[1][3] = 0;
        data[2][0] = axisdata[2] * v1xicosa - axisdata[1] * sina;
        data[2][1] = axisdata[2] * axisdata[1] * icosa + axisdata[0] * sina;
        data[2][2] = cosa + axisdata[2] * axisdata[2] * icosa;
        data[2][3] = 0;
        data[3][0] = 0;
        data[3][1] = 0;
        data[3][2] = 0;
        data[3][3] = 1;
    }

    public static void createTranslationMatrix(Vect3 axis, Matrix4 erg) {
        double data[][] = erg.getData();

        double[] axisdata = axis.getData();

        data[1][0] = 0;
        data[2][0] = 0;
        data[3][0] = 0;
        data[0][1] = 0;
        data[2][1] = 0;
        data[3][1] = 0;
        data[0][2] = 0;
        data[1][2] = 0;
        data[3][2] = 0;

        data[0][0] = 1;
        data[1][1] = 1;
        data[2][2] = 1;
        data[3][3] = 1;
        data[0][3] = axisdata[0];
        data[1][3] = axisdata[1];
        data[2][3] = axisdata[2];
    }

    public static void createScaleMatrix(Vect3 axis, Matrix4 erg) {
        double ergData[][] = erg.getData();
        double axisData[] = axis.getData();

        ergData[0][0] = axisData[0];
        ergData[0][1] = 0;
        ergData[0][2] = 0;
        ergData[0][3] = 0;
        ergData[1][0] = 0;
        ergData[1][1] = axisData[1];
        ergData[1][2] = 0;
        ergData[1][3] = 0;
        ergData[2][0] = 0;
        ergData[2][1] = 0;
        ergData[2][2] = axisData[2];
        ergData[2][3] = 0;
        ergData[3][0] = 0;
        ergData[3][1] = 0;
        ergData[3][2] = 0;
        ergData[3][3] = 1;
    }

    public static void createUnitMatrix(Matrix4 erg) {
        double[][] data = erg.getData();
        data[0][0] = 1;
        data[0][1] = 0;
        data[0][2] = 0;
        data[0][3] = 0;
        data[1][0] = 0;
        data[1][1] = 1;
        data[1][2] = 0;
        data[1][3] = 0;
        data[2][0] = 0;
        data[2][1] = 0;
        data[2][2] = 1;
        data[2][3] = 0;
        data[3][0] = 0;
        data[3][1] = 0;
        data[3][2] = 0;
        data[3][3] = 1;
    }

    public static void invert(Matrix4 m, Matrix4 erg) {
        double[][] a = m.getData();
        double detA = a[0][0] * (a[1][1] * a[2][2] * a[3][3] +
                a[1][2] * a[2][3] * a[3][1] +
                a[1][3] * a[2][1] * a[3][2]) +
                a[0][1] * (a[1][0] * a[2][3] * a[3][2] +
                a[1][2] * a[2][0] * a[3][3] +
                a[1][3] * a[2][2] * a[3][0]) +
                a[0][2] * (a[1][0] * a[2][1] * a[3][3] +
                a[1][1] * a[2][3] * a[3][0] +
                a[1][3] * a[2][0] * a[3][1]) +
                a[0][3] * (a[1][0] * a[2][2] * a[3][1] +
                a[1][1] * a[2][0] * a[3][2] +
                a[1][2] * a[2][1] * a[3][0]) -
                a[0][0] * (a[1][1] * a[2][3] * a[3][2] +
                a[1][2] * a[2][1] * a[3][3] +
                a[1][3] * a[2][2] * a[3][1]) -
                a[0][1] * (a[1][0] * a[2][2] * a[3][3] +
                a[1][2] * a[2][3] * a[3][0] +
                a[1][3] * a[2][0] * a[3][2]) -
                a[0][2] * (a[1][0] * a[2][3] * a[3][1] +
                a[1][1] * a[2][0] * a[3][3] +
                a[1][3] * a[2][1] * a[3][0]) -
                a[0][3] * (a[1][0] * a[2][1] * a[3][2] +
                a[1][1] * a[2][2] * a[3][0] +
                a[1][2] * a[2][0] * a[3][1]);

        detA = 1 / detA;
        double b[][] = erg.getData();
        b[0][0] = (a[1][1] * a[2][2] * a[3][3] + a[1][2] * a[2][3] * a[3][1] + a[1][3] * a[2][1] * a[3][2] - a[1][1] * a[2][3] * a[3][2] - a[1][2] * a[2][1] * a[3][3] - a[1][3] * a[2][2] * a[3][1]) * detA;
        b[0][1] = (a[0][1] * a[2][3] * a[3][2] + a[0][2] * a[2][1] * a[3][3] + a[0][3] * a[2][2] * a[3][1] - a[0][1] * a[2][2] * a[3][3] - a[0][2] * a[2][3] * a[3][1] - a[0][3] * a[2][1] * a[3][2]) * detA;
        b[0][2] = (a[0][1] * a[1][2] * a[3][3] + a[0][2] * a[1][3] * a[3][1] + a[0][3] * a[1][1] * a[3][2] - a[0][1] * a[1][3] * a[3][2] - a[0][2] * a[1][1] * a[3][3] - a[0][3] * a[1][2] * a[3][1]) * detA;
        b[0][3] = (a[0][1] * a[1][3] * a[2][2] + a[0][2] * a[1][1] * a[2][3] + a[0][3] * a[1][2] * a[2][1] - a[0][1] * a[1][2] * a[2][3] - a[0][2] * a[1][3] * a[2][1] - a[0][3] * a[1][1] * a[2][2]) * detA;
        b[1][0] = (a[1][0] * a[2][3] * a[3][2] + a[1][2] * a[2][0] * a[3][3] + a[1][3] * a[2][2] * a[3][0] - a[1][0] * a[2][2] * a[3][3] - a[1][2] * a[2][3] * a[3][0] - a[1][3] * a[2][0] * a[3][2]) * detA;
        b[1][1] = (a[0][0] * a[2][2] * a[3][3] + a[0][2] * a[2][3] * a[3][0] + a[0][3] * a[2][0] * a[3][2] - a[0][0] * a[2][3] * a[3][2] - a[0][2] * a[2][0] * a[3][3] - a[0][3] * a[2][2] * a[3][0]) * detA;
        b[1][2] = (a[0][0] * a[1][3] * a[3][2] + a[0][2] * a[1][0] * a[3][3] + a[0][3] * a[1][2] * a[3][0] - a[0][0] * a[1][2] * a[3][3] - a[0][2] * a[1][3] * a[3][0] - a[0][3] * a[1][0] * a[3][2]) * detA;
        b[1][3] = (a[0][0] * a[1][2] * a[2][3] + a[0][2] * a[1][3] * a[2][0] + a[0][3] * a[1][0] * a[2][2] - a[0][0] * a[1][3] * a[2][2] - a[0][2] * a[1][0] * a[2][3] - a[0][3] * a[1][2] * a[2][0]) * detA;
        b[2][0] = (a[1][0] * a[2][1] * a[3][3] + a[1][1] * a[2][3] * a[3][0] + a[1][3] * a[2][0] * a[3][1] - a[1][0] * a[2][3] * a[3][1] - a[1][1] * a[2][0] * a[3][3] - a[1][3] * a[2][1] * a[3][0]) * detA;
        b[2][1] = (a[0][0] * a[2][3] * a[3][1] + a[0][1] * a[2][0] * a[3][3] + a[0][3] * a[2][1] * a[3][0] - a[0][0] * a[2][1] * a[3][3] - a[0][1] * a[2][3] * a[3][0] - a[0][3] * a[2][0] * a[3][1]) * detA;
        b[2][2] = (a[0][0] * a[1][1] * a[3][3] + a[0][1] * a[1][3] * a[3][0] * a[0][3] * a[1][0] * a[3][1] - a[0][0] * a[1][3] * a[3][1] - a[0][1] * a[1][0] * a[3][3] - a[0][3] * a[1][1] * a[3][0]) * detA;
        b[2][3] = (a[0][0] * a[1][3] * a[2][1] + a[0][1] * a[1][0] * a[2][3] + a[0][3] * a[1][1] * a[2][0] - a[0][0] * a[1][1] * a[2][3] - a[0][1] * a[1][3] * a[2][0] - a[0][3] * a[1][0] * a[2][1]) * detA;
        b[3][0] = (a[1][0] * a[2][2] * a[3][1] + a[1][1] * a[2][0] * a[3][2] + a[1][2] * a[2][1] * a[3][0] - a[1][0] * a[2][1] * a[3][2] - a[1][1] * a[2][2] * a[3][0] - a[1][2] * a[2][0] * a[3][1]) * detA;
        b[3][1] = (a[0][0] * a[2][1] * a[3][2] + a[0][1] * a[2][2] * a[3][0] + a[0][2] * a[2][0] * a[3][1] - a[0][0] * a[2][2] * a[3][1] - a[0][1] * a[2][0] * a[3][2] - a[0][2] * a[2][1] * a[3][0]) * detA;
        b[3][2] = (a[0][0] * a[1][2] * a[3][1] + a[0][1] * a[1][0] * a[3][2] + a[0][2] * a[1][1] * a[3][0] - a[0][0] * a[1][1] * a[3][2] - a[0][1] * a[1][2] * a[3][0] - a[0][2] * a[1][0] * a[3][1]) * detA;
        b[3][3] = (a[0][0] * a[1][1] * a[2][2] + a[0][1] * a[1][2] * a[2][0] + a[0][2] * a[1][0] * a[2][1] - a[0][0] * a[1][2] * a[2][1] - a[0][1] * a[1][0] * a[2][2] - a[0][2] * a[1][1] * a[2][0]) * detA;

    }

    public static void multiply(Matrix4 m1, Matrix4 m2, Matrix4 erg) {
        double h[][] = erg.getData();

        double[][] a = m1.getData();
        double[][] b = m2.getData();

        h[0][0] = a[0][0] * b[0][0] + a[1][0] * b[0][1] + a[2][0] * b[0][2] + a[3][0] * b[0][3];
        h[1][0] = a[0][0] * b[1][0] + a[1][0] * b[1][1] + a[2][0] * b[1][2] + a[3][0] * b[1][3];
        h[2][0] = a[0][0] * b[2][0] + a[1][0] * b[2][1] + a[2][0] * b[2][2] + a[3][0] * b[2][3];
        h[3][0] = a[0][0] * b[3][0] + a[1][0] * b[3][1] + a[2][0] * b[3][2] + a[3][0] * b[3][3];

        h[0][1] = a[0][1] * b[0][0] + a[1][1] * b[0][1] + a[2][1] * b[0][2] + a[3][1] * b[0][3];
        h[1][1] = a[0][1] * b[1][0] + a[1][1] * b[1][1] + a[2][1] * b[1][2] + a[3][1] * b[1][3];
        h[2][1] = a[0][1] * b[2][0] + a[1][1] * b[2][1] + a[2][1] * b[2][2] + a[3][1] * b[2][3];
        h[3][1] = a[0][1] * b[3][0] + a[1][1] * b[3][1] + a[2][1] * b[3][2] + a[3][1] * b[3][3];

        h[0][2] = a[0][2] * b[0][0] + a[1][2] * b[0][1] + a[2][2] * b[0][2] + a[3][2] * b[0][3];
        h[1][2] = a[0][2] * b[1][0] + a[1][2] * b[1][1] + a[2][2] * b[1][2] + a[3][2] * b[1][3];
        h[2][2] = a[0][2] * b[2][0] + a[1][2] * b[2][1] + a[2][2] * b[2][2] + a[3][2] * b[2][3];
        h[3][2] = a[0][2] * b[3][0] + a[1][2] * b[3][1] + a[2][2] * b[3][2] + a[3][2] * b[3][3];

        h[0][3] = a[0][3] * b[0][0] + a[1][3] * b[0][1] + a[2][3] * b[0][2] + a[3][3] * b[0][3];
        h[1][3] = a[0][3] * b[1][0] + a[1][3] * b[1][1] + a[2][3] * b[1][2] + a[3][3] * b[1][3];
        h[2][3] = a[0][3] * b[2][0] + a[1][3] * b[2][1] + a[2][3] * b[2][2] + a[3][3] * b[2][3];
        h[3][3] = a[0][3] * b[3][0] + a[1][3] * b[3][1] + a[2][3] * b[3][2] + a[3][3] * b[3][3];
    }
}

File: JRayXLib/JRayXLib/Math/Matrix.cs
using JRayXLib.Shapes;

namespace JRayXLib.Math
{
    public static class Matrix
    {
        /**
     * creates a rotation matrix for affine transformation
     * @param axis axis to Rotate
     * @param angleRad angle to Rotate
     * @param erg Matrix4 object to store the result
     */

        public static Matrix4 CreateRotationMatrix(Vect3 axis, double angleRad)
        {
            double cosa = System.Math.Cos(angleRad);
            double sina = System.Math.Sin(angleRad);

            double icosa = 1 - cosa; // performance ... every bit matters

            double v1xICosA = axis.X*icosa;

            return new Matrix4
                {
                    A0 = cosa + axis.X*v1xICosA,
                    A1 = axis.Y*v1xICosA - axis.Z*sina,
                    A2 = axis.Z*v1xICosA + axis.Y*sina,
                    B0 = axis.Y*v1xICosA + axis.Z*sina,
                    B1 = cosa + axis.Y*axis.Y*icosa,
                    B2 = axis.Y*axis.Z*icosa - axis.X*sina,
                    C0 = axis.Z*v1xICosA - axis.Y*sina,
                    C1 = axis.Z*axis.Y*icosa + axis.X*sina,
                    C2 = cosa + axis.Z*axis.Z*icosa,
                    D3 = 1
                };
        }

        public static Matrix4 CreateTranslationMatrix(Vect3 axis)
        {
            return new Matrix4
                {
                    A0 = 1,
                    B1 = 1,
                    C2 = 1,
                    D3 = 1,
                    A3 = axis.X,
                    B3 = axis.Y,
                    C3 = axis.Z
                };
        }

        public static Matrix4 CreateScaleMatrix(Vect3 axis)
        {
            return new Matrix4
                {
                    A0 = axis.X,
                    B1 = axis.Y,
                    C2 = axis.Z,
                    D3 = 1
                };
        }

        public static Matrix4 CreateUnitMatrix()
        {
            return new Matrix4
                {
                    A0 = 1,
                    B1 = 1,
                    C2 = 1,
                    D3 = 1
                };
        }

        public static Matrix4 Invert(Matrix4 m)
        {

            double detA = m.A0* ( m.B1 *m.C2 *m.D3 + m.B2 *m.C3 *m.D1 + m.B3 *m.C1 *m.D2 ) +
                          m.A1* ( m.B0 *m.C3 *m.D2 + m.B2 *m.C0 *m.D3 + m.B3 *m.C2 *m.D0 ) +
                          m.A2* ( m.B0 *m.C1 *m.D3 + m.B1 *m.C3 *m.D0 + m.B3 *m.C0 *m.D1 ) +
                          m.A3* ( m.B0 *m.C2 *m.D1 + m.B1 *m.C0 *m.D2 + m.B2 *m.C1 *m.D0 ) -
                          m.A0* ( m.B1 *m.C3 *m.D2 + m.B2 *m.C1 *m.D3 + m.B3 *m.C2 *m.D1 ) -
                          m.A1* ( m.B0 *m.C2 *m.D3 + m.B2 *m.C3 *m.D0 + m.B3 *m.C0 *m.D2 ) -
                          m.A2* ( m.B0 *m.C3 *m.D1 + m.B1 *m.C0 *m.D3 + m.B3 *m.C1 *m.D0 ) -
                          m.A3* ( m.B0 *m.C1 *m.D2 + m.B1 *m.C2 *m.D0 + m.B2 *m.C0 *m.D1 );
            
            detA = 1/detA;

            var erg = new Matrix4
                {
                    A0 = (m.B1*m.C2*m.D3 + m.B2*m.C3*m.D1 + m.B3*m.C1*m.D2 -
                          m.B1*m.C3*m.D2 - m.B2*m.C1*m.D3 - m.B3*m.C2*m.D1)*detA,
                    A1 = (m.A1*m.C3*m.D2 + m.A2*m.C1*m.D3 + m.A3*m.C2*m.D1 -
                          m.A1*m.C2*m.D3 - m.A2*m.C3*m.D1 - m.A3*m.C1*m.D2)*detA,
                    A2 = (m.A1*m.B2*m.D3 + m.A2*m.B3*m.D1 + m.A3*m.B1*m.D2 -
                          m.A1*m.B3*m.D2 - m.A2*m.B1*m.D3 - m.A3*m.B2*m.D1)*detA,
                    A3 = (m.A1*m.B3*m.C2 + m.A2*m.B1*m.C3 + m.A3*m.B2*m.C1 -
                          m.A1*m.B2*m.C3 - m.A2*m.B3*m.C1 - m.A3*m.B1*m.C2)*detA,
                    B0 = (m.B0*m.C3*m.D2 + m.B2*m.C0*m.D3 + m.B3*m.C2*m.D0 -
                          m.B0*m.C2*m.D3 - m.B2*m.C3*m.D0 - m.B3*m.C0*m.D2)*detA,
                    B1 = (m.A0*m.C2*m.D3 + m.A2*m.C3*m.D0 + m.A3*m.C0*m.D2 -
                          m.A0*m.C3*m.D2 - m.A2*m.C0*m.D3 - m.A3*m.C2*m.D0)*detA,
                    B2 = (m.A0*m.B3*m.D2 + m.A2*m.B0*m.D3 + m.A3*m.B2*m.D0 -
                          m.A0*m.B2*m.D3 - m.A2*m.B3*m.D0 - m.A3*m.B0*m.D2)*detA,
                    B3 = (m.A0*m.B2*m.C3 + m.A2*m.B3*m.C0 + m.A3*m.B0*m.C2 -
                          m.A0*m.B3*m.C2 - m.A2*m.B0*m.C3 - m.A3*m.B2*m.C0)*detA,
                    C0 = (m.B0*m.C1*m.D3 + m.B1*m.C3*m.D0 + m.B3*m.C0*m.D1 -
                          m.B0*m.C3*m.D1 - m.B1*m.C0*m.D3 - m.B3*m.C1*m.D0)*detA,
                    C1 = (m.A0*m.C3*m.D1 + m.A1*m.C0*m.D3 + m.A3*m.C1*m.D0 -
                          m.A0*m.C1*m.D3 - m.A1*m.C3*m.D0 - m.A3*m.C0*m.D1)*detA,
                    C2 = (m.A0*m.B1*m.D3 + m.A1*m.B3*m.D0 + m.A3*m.B0*m.D1 -
                          m.A0*m.B3*m.D1 - m.A1*m.B0*m.D3 - m.A3*m.B1*m.D0)*detA,
                    C3 = (m.A0*m.B3*m.C1 + m.A1*m.B0*m.C3 + m.A3*m.B1*m.C0 -
                          m.A0*m.B1*m.C3 - m.A1*m.B3*m.C0 - m.A3*m.B0*m.C1)*detA,
                    D0 = (m.B0*m.C2*m.D1 + m.B1*m.C0*m.D2 + m.B2*m.C1*m.D0 -
                          m.B0*m.C1*m.D2 - m.B1*m.C2*m.D0 - m.B2*m.C0*m.D1)*detA,
                    D1 = (m.A0*m.C1*m.D2 + m.A1*m.C2*m.D0 + m.A2*m.C0*m.D1 -
                          m.A0*m.C2*m.D1 - m.A1*m.C0*m.D2 - m.A2*m.C1*m.D0)*detA,
                    D2 = (m.A0*m.B2*m.D1 + m.A1*m.B0*m.D2 + m.A2*m.B1*m.D0 -
                          m.A0*m.B1*m.D2 - m.A1*m.B2*m.D0 - m.A2*m.B0*m.D1)*detA,
                    D3 = (m.A0*m.B1*m.C2 + m.A1*m.B2*m.C0 + m.A2*m.B0*m.C1 -
                          m.A0*m.B2*m.C1 - m.A1*m.B0*m.C2 - m.A2*m.B1*m.C0)*detA
                };

            return erg;
        }

        public static Matrix4 Multiply(Matrix4 m1, Matrix4 m2)
        {
            return new Matrix4
                {
                    A0 = m1.A0*m2.A0 + m1.B0*m2.A1 + m1.C0*m2.A2 + m1.D0*m2.A3,
                    B0 = m1.A0*m2.B0 + m1.B0*m2.B1 + m1.C0*m2.B2 + m1.D0*m2.B3,
                    C0 = m1.A0*m2.C0 + m1.B0*m2.C1 + m1.C0*m2.C2 + m1.D0*m2.C3,
                    D0 = m1.A0*m2.D0 + m1.B0*m2.D1 + m1.C0*m2.D2 + m1.D0*m2.D3,
                    A1 = m1.A1*m2.A0 + m1.B1*m2.A1 + m1.C1*m2.A2 + m1.D1*m2.A3,
                    B1 = m1.A1*m2.B0 + m1.B1*m2.B1 + m1.C1*m2.B2 + m1.D1*m2.B3,
                    C1 = m1.A1*m2.C0 + m1.B1*m2.C1 + m1.C1*m2.C2 + m1.D1*m2.C3,
                    D1 = m1.A1*m2.D0 + m1.B1*m2.D1 + m1.C1*m2.D2 + m1.D1*m2.D3,
                    A2 = m1.A2*m2.A0 + m1.B2*m2.A1 + m1.C2*m2.A2 + m1.D2*m2.A3,
                    B2 = m1.A2*m2.B0 + m1.B2*m2.B1 + m1.C2*m2.B2 + m1.D2*m2.B3,
                    C2 = m1.A2*m2.C0 + m1.B2*m2.C1 + m1.C2*m2.C2 + m1.D2*m2.C3,
                    D2 = m1.A2*m2.D0 + m1.B2*m2.D1 + m1.C2*m2.D2 + m1.D2*m2.D3,
                    A3 = m1.A3*m2.A0 + m1.B3*m2.A1 + m1.C3*m2.A2 + m1.D3*m2.A3,
                    B3 = m1.A3*m2.B0 + m1.B3*m2.B1 + m1.C3*m2.B2 + m1.D3*m2.B3,
                    C3 = m1.A3*m2.C0 + m1.B3*m2.C1 + m1.C3*m2.C2 + m1.D3*m2.C3,
                    D3 = m1.A3*m2.D0 + m1.B3*m2.D1 + m1.C3*m2.D2 + m1.D3*m2.D3
                };
        }
    }
}